subsys/emds/emds_new.c

```diff
@@ -254,21 +258,83 @@ int emds_load(void)
 	}
 
 	LOG_DBG("Found freshest snapshot in partition %d with fresh_cnt %u",
-		freshest_partition_idx, freshest_snapshot.fresh_cnt);
+		freshest_partition_idx, freshest_snapshot.metadata.fresh_cnt);
+
+	rc = emds_read_data(partition[freshest_partition_idx].fa, &freshest_snapshot.metadata);
+	if (rc == 0) {
+		emds_state = EMDS_STATE_SYNCHRONIZED;
+	}
 
-	return emds_read_data(partition[freshest_partition_idx].fa, &freshest_snapshot);
+	return rc;
 }
 
 int emds_prepare(void)
 {
-	emds_ready = true;
+	size_t data_size;
+	bool erase_enabled = false;
+	int idx;
+	int freshest_partition_idx = -1;
+	int rc = 0;
 
-	return 0;
+	if (emds_state != EMDS_STATE_SYNCHRONIZED) {
+		return -ECANCELED;
+	}
+
+	/* Returned status is not checked since initialization state is checked above */
+	(void)emds_store_size_get(&data_size);
+
+	/** First try to allocate snapshot in the same partition where freshest snapshot exists */
+	if (freshest_snapshot.metadata.fresh_cnt > 0) {
+		for (idx = 0; idx < PARTITIONS_NUM_MAX; idx++) {
+			if (IN_RANGE(freshest_snapshot.metadata_addr, partition[idx].fa->fa_off,
+				     partition[idx].fa->fa_off + partition[idx].fa->fa_size)) {
+				freshest_partition_idx = idx;
+				break;
+			}
+		}
+
+		rc = emds_flash_allocate_snapshot(&partition[idx], &freshest_snapshot,
+						  &allocated_snapshot, data_size);
+		if (rc == 0) {
+			emds_state = EMDS_STATE_READY;
+			return 0;
+		}
+	}
+
+	idx = -1;
+	do {
+		idx++;
+
+		if (idx != freshest_partition_idx) {
+			if (erase_enabled) {
+				LOG_DBG("Erase partition %d", idx);
+				rc = emds_flash_erase_partition(&partition[idx]);
+				if (rc) {
+					LOG_ERR("Failed to erase partition %d: %d", idx, rc);
+					break;
+				}
+			}
+
+			rc = emds_flash_allocate_snapshot(&partition[idx], NULL,
+							  &allocated_snapshot, data_size);
+			if (rc == 0) {
+				emds_state = EMDS_STATE_READY;
+				return 0;
+			}
+		}
+
+		if ((idx == PARTITIONS_NUM_MAX - 1) && !erase_enabled && rc == -EADDRINUSE) {
```

Comment:
Note: `rc == -EADDRINUSE` could happen due to result of rc from line 296. I hope this is intentional.